### PR TITLE
Constrain raiseError with ApplicativeError

### PR DESCRIPTION
**Pull request by bpholt**
I [asked on gitter](https://gitter.im/functional-streams-for-scala/fs2?at=5b74a2b049932d4fe4e8ac65) whether there was any `Either[Throwable, A] ⇒ Stream[Pure, A]` method in fs2. After some discussion, the conclusion that it doesn't make sense to have a `Pure` version, but a `Stream.fromEither` constrained to an `F[_]` with an `ApplicativeError[F, Throwable]` was a good idea.

This PR is to constrain the methods already in fs2 in the same way, so we can add `Stream.fromEither`, either in an addition to this PR or in a subsequent PR.

Thanks to @SystemFw, @mpilquist, and @ChristopherDavenport for helping me figure out how to make the necessary changes and tweaking the initial question into something worth adding to fs2.

### CODE REVIEW

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 1]
```diff
@@ -2452,16 +2453,16 @@ object Stream {
...
-    * scala> Try(Stream.raiseError(new RuntimeException).toList)
-    * res0: Try[List[Nothing]] = Failure(java.lang.RuntimeException)
     * scala> import cats.effect.IO
-    * scala> Stream.raiseError(new RuntimeException).covary[IO].compile.drain.attempt.unsafeRunSync
+    * scala> Stream.raiseError[IO, Nothing](new RuntimeException).compile.drain.attempt.unsafeRunSync
     * res0: Either[Throwable,Unit] = Left(java.lang.RuntimeException)
     * }}}
     */
-  def raiseError[F[x] >: Pure[x], O](e: Throwable): Stream[F, O] =
+  def raiseError[F[x], O](e: Throwable)(
```
Comment (bpholt) [inline core/shared/src/main/scala/fs2/Stream.scala]: When I was adding the explicit types to calls to `raiseError` I noticed that I was able to use `Nothing` as the output type pretty much everywhere. That made me wonder if the signature of this method should change to 

```scala
def raiseError[F[x]](e: Throwable)(implicit AE: ApplicativeError[F, Throwable]): Stream[F, Nothing]
```

which I think would be a little easier to deal with as a caller?

I left 77f9d0ba as a separate commit for now to show the difference, but I assume we'll either want to revert it or squash it, depending on whether sticking with `Nothing` makes sense.

Comment (mpilquist) [reply core/shared/src/main/scala/fs2/Stream.scala]: Yeah, let's remove the output param from `raiseError`.

Comment (bpholt) [reply core/shared/src/main/scala/fs2/Stream.scala]: Implemented and squashed in 3a64d4d4

**Comment by bpholt** [comment]
FYI, I added `Stream.fromEither` in 7d16e7f

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 2]
```diff
@@ -3351,3 +3373,19 @@ object Stream {
       def combine(x: Stream[F, O], y: Stream[F, O]) = x ++ y
     }
 }
+
+trait StreamLowPriority {
```
Comment (mpilquist) [inline core/shared/src/main/scala/fs2/Stream.scala]: Make this trait `private[fs2]`

Comment (bpholt) [reply core/shared/src/main/scala/fs2/Stream.scala]: Updated in 1389078b

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 3]
```diff
@@ -3073,7 +3075,7 @@ object Stream {
           case None => Pull.pure(None)
           case Some((hd, tl)) =>
             hd.size.toLong match {
-              case m if m < n => Pull.output(hd) >> tl.pull.take(n - m)
+              case m if m < n  => Pull.output(hd) >> tl.pull.take(n - m)
```
[the remark this answers is not in the document]
Comment (bpholt) [reply core/shared/src/main/scala/fs2/Stream.scala]: I ran `sbt clean +test` and committed the formatting changes it made prior to making the PR. Is there something else that should have been run too?

[the remark this answers is not in the document]
Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/Stream.scala]: I don't care that much formatting, is just strange sudently space is added :-) Should be ok if you run tests b4 commit.

File: core/shared/src/main/scala/fs2/Stream.scala [hunk 4]
```diff
@@ -1371,7 +1371,7 @@ final class Stream[+F[_], +O] private (private val free: FreeC[Algebra[Nothing,
     * }}}
     */
   def onComplete[F2[x] >: F[x], O2 >: O](s2: => Stream[F2, O2]): Stream[F2, O2] =
-    handleErrorWith(e => s2 ++ Stream.raiseError(e)) ++ s2
+    handleErrorWith(e => s2 ++ Stream.fromFreeC(Algebra.raiseError(e))) ++ s2
```
[the remark this answers is not in the document]
Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/Stream.scala]: Thank you. 

@mpilquist @SystemFw I am sort of missing the benefits of this. I really liked that Streams _did not_ require any Monad/Applicative Errors. Can you please give me short hint on this?

File: core/shared/src/main/scala/fs2/Stream.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (bpholt) [reply core/shared/src/main/scala/fs2/Stream.scala]: FWIW, I confirmed the space is added by the formatter. It appears to be aligning the `=>` arrows across the different cases.

File: core/shared/src/main/scala/fs2/Stream.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (SystemFw) [reply core/shared/src/main/scala/fs2/Stream.scala]: Just woke up :)

```scala
def well : Stream[Pure, Int] = Stream(1,2,3) ++ Stream.raiseError(new Exception("Booom!"))

// I think I'm in safe pure land
val a = well.toList
// new java.lang.Exception("Booom!")
```

[the remark this answers is not in the document]
Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/Stream.scala]: @SystemFw makes sense. It adds machinary, but now I completely understand :-) .Thanks

File: core/shared/src/main/scala/fs2/Stream.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (pchlupacek) [reply core/shared/src/main/scala/fs2/Stream.scala]: cool. thanks.

File: core/shared/src/main/scala/fs2/Stream.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (SystemFw) [reply core/shared/src/main/scala/fs2/Stream.scala]: Yes, I also had doubts wrt machinery but it turns out it's not _too_ bad, so we went ahead :)

**Pull request merged**